File: ts-sdk/client/src/state/tickArray.ts
import {
  Account,
  Address,
  assertAccountExists,
  assertAccountsExist,
  EncodedAccount,
  Encoder,
  FetchAccountConfig,
  FetchAccountsConfig,
  fetchEncodedAccount,
  fetchEncodedAccounts,
  fixEncoderSize,
  getAddressEncoder,
  getArrayEncoder,
  getBytesEncoder,
  getI32Encoder,
  getStructEncoder,
  getU64Decoder,
  MaybeAccount,
  MaybeEncodedAccount,
  ReadonlyUint8Array,
  transformEncoder,
} from "@solana/kit";

import {
  decodeFixedTickArray,
  decodeSparseTickArray,
  FIXED_TICK_ARRAY_DISCRIMINATOR,
  FixedTickArray,
  getMaybeTickEncoder,
  MaybeTick,
  MaybeTickArgs,
  SPARSE_TICK_ARRAY_DISCRIMINATOR,
  SparseTickArray,
  Tick,
  type TickArgs,
} from "../generated";

export type TickArray = {
  discriminator: ReadonlyUint8Array;
  startTickIndex: number;
  fusionPool: Address;
  ticks: Array<Tick>;
};

export type TickArrayArgs = {
  startTickIndex: number;
  fusionPool: Address;
  ticks: Array<TickArgs>;
};

const FIXED_TICK_ARRAY_DISCRIMINATOR_NUMBER = getU64Decoder().decode(FIXED_TICK_ARRAY_DISCRIMINATOR);
const SPARSE_TICK_ARRAY_DISCRIMINATOR_NUMBER = getU64Decoder().decode(SPARSE_TICK_ARRAY_DISCRIMINATOR);

export function getTickArrayMinSize(): number {
  return 132; // 8+4+32+88
}

export function getTickArrayMaxSize(): number {
  return 9988; // 8+4+32+88*113
}

export function consolidateTick(tick: Tick | MaybeTick): Tick {
  if ("initialized" in tick) {
    return tick;
  }
  switch (tick.__kind) {
    case "Uninitialized":
      return {
        initialized: false,
        liquidityGross: 0n,
        liquidityNet: 0n,
        feeGrowthOutsideA: 0n,
        feeGrowthOutsideB: 0n,
        age: 0n,
        fulfilledAToBOrdersInput: 0n,
        fulfilledBToAOrdersInput: 0n,
        openOrdersInput: 0n,
        partFilledOrdersInput: 0n,
        partFilledOrdersRemainingInput: 0n,
      };
    case "Initialized":
      return {
        initialized: true,
        liquidityGross: tick.fields[0].liquidityGross,
        liquidityNet: tick.fields[0].liquidityNet,
        feeGrowthOutsideA: tick.fields[0].feeGrowthOutsideA,
        feeGrowthOutsideB: tick.fields[0].feeGrowthOutsideB,
        age: tick.fields[0].age,
        fulfilledAToBOrdersInput: tick.fields[0].fulfilledAToBOrdersInput,
        fulfilledBToAOrdersInput: tick.fields[0].fulfilledBToAOrdersInput,
        openOrdersInput: tick.fields[0].openOrdersInput,
        partFilledOrdersInput: tick.fields[0].partFilledOrdersInput,
        partFilledOrdersRemainingInput: tick.fields[0].partFilledOrdersRemainingInput,
      };
  }
}

export function getTickArrayEncoder(): Encoder<TickArrayArgs> {
  return transformEncoder(
    getStructEncoder([
      ["discriminator", fixEncoderSize(getBytesEncoder(), 8)],
      ["startTickIndex", getI32Encoder()],
      ["fusionPool", getAddressEncoder()],
      ["ticks", getArrayEncoder(getMaybeTickEncoder(), { size: 88 })],
    ]),
    value => ({
      discriminator: SPARSE_TICK_ARRAY_DISCRIMINATOR,
      startTickIndex: value.startTickIndex,
      fusionPool: value.fusionPool,
      ticks: value.ticks.map((t): MaybeTickArgs => {
        if (t.initialized) {
          return {
            __kind: "Initialized",
            fields: [
              {
                liquidityNet: t.liquidityNet,
                liquidityGross: t.liquidityGross,
                feeGrowthOutsideA: t.feeGrowthOutsideA,
                feeGrowthOutsideB: t.feeGrowthOutsideB,
                age: t.age,
                openOrdersInput: t.openOrdersInput,
                partFilledOrdersInput: t.partFilledOrdersInput,
                partFilledOrdersRemainingInput: t.partFilledOrdersRemainingInput,
                fulfilledAToBOrdersInput: t.fulfilledAToBOrdersInput,
                fulfilledBToAOrdersInput: t.fulfilledBToAOrdersInput,
              },
            ],
          };
        } else {
          return {
            __kind: "Uninitialized",
          };
        }
      }),
    }),
  );
}

export function consolidateTickArray<TAddress extends string = string>(
  tickArrayAccount: Account<FixedTickArray | SparseTickArray, TAddress>,
): Account<TickArray, TAddress>;
export function consolidateTickArray<TAddress extends string = string>(
  tickArrayAccount: MaybeAccount<FixedTickArray | SparseTickArray, TAddress>,
): MaybeAccount<TickArray, TAddress>;
export function consolidateTickArray<TAddress extends string = string>(
  tickArrayAccount:
    | Account<FixedTickArray | SparseTickArray, TAddress>
    | MaybeAccount<FixedTickArray | SparseTickArray, TAddress>,
): Account<TickArray, TAddress> | MaybeAccount<TickArray, TAddress> {
  if ("exists" in tickArrayAccount && !tickArrayAccount.exists) {
    return tickArrayAccount;
  }

  return {
    ...tickArrayAccount,
    data: {
      ...tickArrayAccount.data,
      ticks: tickArrayAccount.data.ticks.map(consolidateTick),
    },
  };
}

export function decodeTickArray<TAddress extends string = string>(
  encodedAccount: EncodedAccount<TAddress>,
): Account<TickArray, TAddress>;
export function decodeTickArray<TAddress extends string = string>(
  encodedAccount: MaybeEncodedAccount<TAddress>,
): MaybeAccount<TickArray, TAddress>;
export function decodeTickArray<TAddress extends string = string>(
  encodedAccount: EncodedAccount<TAddress> | MaybeEncodedAccount<TAddress>,
): Account<TickArray, TAddress> | MaybeAccount<TickArray, TAddress> {
  if ("exists" in encodedAccount && !encodedAccount.exists) {
    return encodedAccount;
  }
  const discriminator = getU64Decoder().decode(encodedAccount.data.subarray(0, 8));
  switch (discriminator) {
    case FIXED_TICK_ARRAY_DISCRIMINATOR_NUMBER:
      return consolidateTickArray(decodeFixedTickArray(encodedAccount));
    case SPARSE_TICK_ARRAY_DISCRIMINATOR_NUMBER:
      return consolidateTickArray(decodeSparseTickArray(encodedAccount));
    default:
      throw new Error(`Unknown discriminator: ${discriminator}`);
  }
}

export async function fetchTickArray<TAddress extends string = string>(
  rpc: Parameters<typeof fetchEncodedAccount>[0],
  address: Address<TAddress>,
  config?: FetchAccountConfig,
): Promise<Account<TickArray, TAddress>> {
  const maybeAccount = await fetchMaybeTickArray(rpc, address, config);
  assertAccountExists(maybeAccount);
  return maybeAccount;
}

export async function fetchMaybeTickArray<TAddress extends string = string>(
  rpc: Parameters<typeof fetchEncodedAccount>[0],
  address: Address<TAddress>,
  config?: FetchAccountConfig,
): Promise<MaybeAccount<TickArray, TAddress>> {
  const maybeAccount = await fetchEncodedAccount(rpc, address, config);
  return decodeTickArray(maybeAccount);
}

export async function fetchAllTickArray(
  rpc: Parameters<typeof fetchEncodedAccounts>[0],
  addresses: Array<Address>,
  config?: FetchAccountsConfig,
): Promise<Account<TickArray>[]> {
  const maybeAccounts = await fetchAllMaybeTickArray(rpc, addresses, config);
  assertAccountsExist(maybeAccounts);
  return maybeAccounts;
}

export async function fetchAllMaybeTickArray(
  rpc: Parameters<typeof fetchEncodedAccounts>[0],
  addresses: Array<Address>,
  config?: FetchAccountsConfig,
): Promise<MaybeAccount<TickArray>[]> {
  const maybeAccounts = await fetchEncodedAccounts(rpc, addresses, config);
  return maybeAccounts.map(maybeAccount => decodeTickArray(maybeAccount));
}

File: rust-sdk/client/src/state/tick_array.rs
use crate::{FixedTickArray, MaybeTick, SparseTickArray, Tick, FIXED_TICK_ARRAY_DISCRIMINATOR, SPARSE_TICK_ARRAY_DISCRIMINATOR};
use borsh::{BorshDeserialize, BorshSerialize};
use solana_account_info::AccountInfo;
use solana_pubkey::Pubkey;

pub const TICK_ARRAY_DISCRIMINATOR: [u8; 8] = SPARSE_TICK_ARRAY_DISCRIMINATOR;

#[derive(Clone, Debug, Eq, PartialEq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct TickArray {
    pub discriminator: [u8; 8],
    pub start_tick_index: i32,
    #[cfg_attr(feature = "serde", serde(with = "serde_with::As::<serde_with::DisplayFromStr>"))]
    pub fusion_pool: Pubkey,
    #[cfg_attr(feature = "serde", serde(with = "serde_big_array::BigArray"))]
    pub ticks: [Tick; 88],
}

impl TickArray {
    pub const MIN_LEN: usize = 132; // 8+4+32+88
    pub const MAX_LEN: usize = 9988; // 8+4+32+88*113

    pub fn from_bytes(bytes: &[u8]) -> Result<Self, std::io::Error> {
        if bytes.len() < 8 {
            return Err(std::io::Error::new(std::io::ErrorKind::InvalidInput, "Invalid account data length"));
        }
        let discriminator = &bytes[0..8];
        if discriminator == FIXED_TICK_ARRAY_DISCRIMINATOR {
            let tick_array = FixedTickArray::from_bytes(bytes)?;
            Ok(tick_array.into())
        } else if discriminator == SPARSE_TICK_ARRAY_DISCRIMINATOR {
            let tick_array = SparseTickArray::from_bytes(bytes)?;
            Ok(tick_array.into())
        } else {
            Err(std::io::Error::new(std::io::ErrorKind::InvalidInput, "Invalid account discriminator"))
        }
    }
}

impl From<FixedTickArray> for TickArray {
    fn from(val: FixedTickArray) -> Self {
        TickArray {
            discriminator: TICK_ARRAY_DISCRIMINATOR,
            start_tick_index: val.start_tick_index,
            fusion_pool: val.fusion_pool,
            ticks: val.ticks,
        }
    }
}

impl From<SparseTickArray> for TickArray {
    fn from(val: SparseTickArray) -> Self {
        TickArray {
            discriminator: TICK_ARRAY_DISCRIMINATOR,
            start_tick_index: val.start_tick_index,
            fusion_pool: val.fusion_pool,
            ticks: val.ticks.map(|tick| tick.into()),
        }
    }
}

impl<'a> TryFrom<&AccountInfo<'a>> for TickArray {
    type Error = std::io::Error;

    fn try_from(account_info: &AccountInfo<'a>) -> Result<Self, Self::Error> {
        let data: &[u8] = &(*account_info.data).borrow();
        Self::from_bytes(data)
    }
}

impl From<TickArray> for SparseTickArray {
    fn from(val: TickArray) -> Self {
        SparseTickArray {
            discriminator: SPARSE_TICK_ARRAY_DISCRIMINATOR,
            start_tick_index: val.start_tick_index,
            fusion_pool: val.fusion_pool,
            ticks: val.ticks.map(|tick| tick.into()),
        }
    }
}

impl From<MaybeTick> for Tick {
    fn from(val: MaybeTick) -> Self {
        match val {
            MaybeTick::Initialized(tick) => Tick {
                initialized: true,
                liquidity_net: tick.liquidity_net,
                liquidity_gross: tick.liquidity_gross,
                fee_growth_outside_a: tick.fee_growth_outside_a,
                fee_growth_outside_b: tick.fee_growth_outside_b,
                age: tick.age,
                open_orders_input: tick.open_orders_input,
                part_filled_orders_input: tick.part_filled_orders_input,
                part_filled_orders_remaining_input: tick.part_filled_orders_remaining_input,
                fulfilled_a_to_b_orders_input: tick.fulfilled_a_to_b_orders_input,
                fulfilled_b_to_a_orders_input: tick.fulfilled_b_to_a_orders_input,
            },
            MaybeTick::Uninitialized => Tick {
                initialized: false,
                liquidity_net: 0,
                liquidity_gross: 0,
                fee_growth_outside_a: 0,
                fee_growth_outside_b: 0,
                age: 0,
                open_orders_input: 0,
                part_filled_orders_input: 0,
                part_filled_orders_remaining_input: 0,
                fulfilled_a_to_b_orders_input: 0,
                fulfilled_b_to_a_orders_input: 0,
            },
        }
    }
}

impl BorshSerialize for TickArray {
    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
        SparseTickArray::from(self.clone()).serialize(writer)
    }
}

impl BorshDeserialize for TickArray {
    fn deserialize_reader<R: std::io::Read>(reader: &mut R) -> std::io::Result<Self> {
        let mut buf = Vec::new();
        reader.read_to_end(&mut buf)?;
        Self::from_bytes(&buf)
    }
}

#[cfg(feature = "anchor")]
impl anchor_lang::AccountDeserialize for TickArray {
    fn try_deserialize_unchecked(buf: &mut &[u8]) -> anchor_lang::Result<Self> {
        Ok(Self::from_bytes(buf)?)
    }
}

#[cfg(feature = "anchor")]
impl anchor_lang::AccountSerialize for TickArray {}

#[cfg(feature = "anchor")]
impl anchor_lang::Owner for TickArray {
    fn owner() -> Pubkey {
        crate::FUSIONAMM_ID
    }
}

#[cfg(feature = "anchor-idl-build")]
impl anchor_lang::IdlBuild for TickArray {}

#[cfg(feature = "anchor-idl-build")]
impl anchor_lang::Discriminator for TickArray {
    const DISCRIMINATOR: &[u8] = &[0; 8];
}

#[cfg(feature = "fetch")]
pub fn fetch_tick_array(
    rpc: &solana_client::rpc_client::RpcClient,
    address: &solana_program::pubkey::Pubkey,
) -> Result<crate::DecodedAccount<TickArray>, std::io::Error> {
    let accounts = fetch_all_tick_array(rpc, &[*address])?;
    Ok(accounts[0].clone())
}

#[cfg(feature = "fetch")]
pub fn fetch_all_tick_array(
    rpc: &solana_client::rpc_client::RpcClient,
    addresses: &[solana_program::pubkey::Pubkey],
) -> Result<Vec<crate::shared::DecodedAccount<TickArray>>, std::io::Error> {
    let accounts = rpc
        .get_multiple_accounts(addresses)
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e.to_string()))?;
    let mut decoded_accounts: Vec<crate::shared::DecodedAccount<TickArray>> = Vec::new();
    for i in 0..addresses.len() {
        let address = addresses[i];
        let account = accounts[i]
            .as_ref()
            .ok_or(std::io::Error::new(std::io::ErrorKind::Other, format!("Account not found: {}", address)))?;
        let data = TickArray::from_bytes(&account.data)?;
        decoded_accounts.push(crate::shared::DecodedAccount {
            address,
            account: account.clone(),
            data,
        });
    }
    Ok(decoded_accounts)
}

#[cfg(feature = "fetch")]
pub fn fetch_maybe_tick_array(
    rpc: &solana_client::rpc_client::RpcClient,
    address: &solana_program::pubkey::Pubkey,
) -> Result<crate::shared::MaybeAccount<TickArray>, std::io::Error> {
    let accounts = fetch_all_maybe_tick_array(rpc, &[*address])?;
    Ok(accounts[0].clone())
}

#[cfg(feature = "fetch")]
pub fn fetch_all_maybe_tick_array(
    rpc: &solana_client::rpc_client::RpcClient,
    addresses: &[solana_program::pubkey::Pubkey],
) -> Result<Vec<crate::shared::MaybeAccount<TickArray>>, std::io::Error> {
    let accounts = rpc
        .get_multiple_accounts(addresses)
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e.to_string()))?;
    let mut decoded_accounts: Vec<crate::shared::MaybeAccount<TickArray>> = Vec::new();
    for i in 0..addresses.len() {
        let address = addresses[i];
        if let Some(account) = accounts[i].as_ref() {
            let data = TickArray::from_bytes(&account.data)?;
            decoded_accounts.push(crate::shared::MaybeAccount::Exists(crate::shared::DecodedAccount {
                address,
                account: account.clone(),
                data,
            }));
        } else {
            decoded_accounts.push(crate::shared::MaybeAccount::NotFound(address));
        }
    }
    Ok(decoded_accounts)
}
